metadata/metadata_test.go

```diff
@@ -0,0 +1,181 @@
+			err:      false,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0", "-metadata", `{"example": metadata"}`},
+			metadata: ``,
+			err:      true,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0", "-metadata", `"foobar"`},
+			metadata: `"foobar"`,
+			err:      false,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0", "-metadata", `foobar`},
+			metadata: ``,
+			err:      true,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0", "-metadata", "262144"},
+			metadata: `262144`,
+			err:      false,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0", "-metadata", "true"},
+			metadata: `true`,
+			err:      false,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0", "-metadata", "null"},
+			metadata: `null`,
+			err:      false,
+		},
+		{
+			command:  []string{"go", "run", "testdata/json.go", "-exit-code", "0"},
+			metadata: ``,
+			err:      true,
+		},
+	}
+	for _, test := range tests {
+		g := Generator{
+			Config: &config.MetadataPlugin{
+				CommandArgs: test.command,
+			},
+		}
+		metadata, err := g.Fetch()
+		if err != nil {
+			if !test.err {
+				t.Errorf("error occurred unexpectedly on command %v %s", test.command, err.Error())
+			}
+		} else {
+			if test.err {
+				t.Errorf("error did not occurr but error expected on command %v", test.command)
+			}
+			metadataStr, _ := json.Marshal(metadata)
+			if string(metadataStr) != test.metadata {
+				t.Errorf("metadata should be %q but got %q", test.metadata, string(metadataStr))
+			}
+		}
+	}
+}
+
+func TestMetadataGeneratorSaveDiffers(t *testing.T) {
+	tests := []struct {
+		prevmetadata string
+		metadata     string
+		differs      bool
+	}{
+		{
+			prevmetadata: `{}`,
+			metadata:     `{}`,
+			differs:      false,
+		},
+		{
+			prevmetadata: `{ "foo": [ 100, 200, null, {} ] }`,
+			metadata:     `{"foo":[100,200,null,{}]}`,
+			differs:      false,
+		},
+		{
+			prevmetadata: `null`,
+			metadata:     `{}`,
+			differs:      true,
+		},
+		{
+			prevmetadata: `[]`,
+			metadata:     `{}`,
+			differs:      true,
+		},
+	}
+	for i, test := range tests {
+		g := Generator{Tempfile: "testdata/.mackerel-metadata-test-" + strconv.Itoa(i)}
+		var prevmetadata interface{}
+		_ = json.Unmarshal([]byte(test.prevmetadata), &prevmetadata)
+
+		if err := g.Save(prevmetadata); err != nil {
+			t.Errorf("Error should not occur in Save() but got: %s", err.Error())
+		}
+
+		var metadata interface{}
+		_ = json.Unmarshal([]byte(test.metadata), &metadata)
+
+		got := g.Differs(metadata)
+		if got != test.differs {
+			t.Errorf("Differs() should return %t but got %t for %v, %v", test.differs, got, prevmetadata, metadata)
+		}
+
+		if err := g.Clear(); err != nil {
+			t.Errorf("Error should not occur in Clear() but got: %s", err.Error())
```

Comment:
[nit] it's common to use '%v' to format error, so you don't have to call `err.Error()` and use '%s' everywhere.